### PR TITLE
Adding Container package

### PR DESCRIPTION
**Pull request by lablnet**
#194

### CODE REVIEW

File: src/Container/Container.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
+<?php
```
Comment (peter279k) [inline src/Container/Container.php]: The improved `Container` class need to have the unit tests.

We have to make sure that this `Container` class should be worked as our expected behavior.

Comment (lablnet) [reply src/Container/Container.php]: yes

**Comment by lablnet** [comment]
@peter279k  check the latest commit and add `Tests` file in your available/free time

File: src/Container/Container.php [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+	/**
+	 * Parse the hint parameter.
+	 *
+	 * @param  string|array $hint Type hint or array contaning both type hint and alias
+	 *
+	 * @since 3.0.0
+	 *
+	 * @return string
+	 */
+	protected function parseHint($hint): string
```
Comment (peter279k) [inline src/Container/Container.php]: It looks like the `$hint` will be the `array` or `string`.

Please check the type. I think it should throw `InvalidArgumentException` if `$hint` type is invalid :).

File: src/Container/Container.php [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+	 * @param string|\Closure $class     Class name or closure.
+	 * @param bool            $singleton Should we return the same instance every time?
+	 *
+	 * @since 3.0.0
+	 *
+	 * @return void
+	 */
+	public function register($hint, $class, bool $singleton = false)
+	{
+		$this->hints[$this->parseHint($hint)] = ['class' => $class, 'singleton' => $singleton];
```
Comment (peter279k) [inline src/Container/Container.php]: Please check the `$hint` and `$class` type before assigning them to `$this->hints` array.

File: src/Container/Container.php [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+	 * Creates a class instance using closure.
+	 *
+	 * @param closure $factory      Closuare.
+	 * @param array   $parameters Constructor parameters.
+	 *
+	 * @since 3.0.0
+	 *
+	 * @return object
+	 */
+	public function closureFactory(\Closure $factory, $parameters)
```
Comment (peter279k) [inline src/Container/Container.php]: It should add the `array` type hint for `$parameters`.

File: src/Contracts/Container/Container.php [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+     *
+     * @param string $class          Class.
+     * @param string $interface      Interface.
+     * @param string $implementation Implementation.
+     *
+     * @since 3.0.0
+     *
+     * @return void
+     */
+    public function registerContextualDependency($class, string $interface, string $implementation);
```
Comment (peter279k) [inline src/Contracts/Container/Container.php:69]: I think it should add the `sting` type hint for `$class`.

File: src/Contracts/Container/Container.php [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+     * Creates a class instance using closure.
+     *
+     * @param closure $factory      Closuare.
+     * @param array   $parameters Constructor parameters.
+     *
+     * @since 3.0.0
+     *
+     * @return object
+     */
+    public function closureFactory(\Closure $factory, $parameters);
```
Comment (peter279k) [inline src/Contracts/Container/Container.php]: Add the `array` type hint for `$parameters`.

**Review by peter279k — CHANGES_REQUESTED** [review]
Please consider the parameter type hints for all methods.

I write the comments on that line and note that considering type hints.

Some places that don't have comments, but it's similar problem on other comments.

You should check all argument type hints on methods :).

File: src/Container/Container.php (hunk 3 above)
Comment (lablnet) [reply src/Container/Container.php]: Yes `$hint` will check in `parseHint` method
could you explain how we check the class parameter?

File: src/Container/Container.php (hunk 2 above)
Comment (lablnet) [reply src/Container/Container.php]: Yes, consider the following code
```php
	protected function parseHint($hint): string
	{
		if(Arrays::isReallyArray($hint))
		{
			list($hint, $alias) = $hint;

			$this->aliases[$alias] = $hint;
		} elseif (!preg_match("/a-zA-Z/i", $hint)) {
			throw new \InvalidArgumentException ("The {$hint} parameter should be array or string ".gettype($hind)." given", 500);
			
		}

		return $hint;
	}
```

Comment (peter279k) [reply src/Container/Container.php]: @Lablnet, thank you for your reply.

This code snippets have following problems:

- coding style is about `if` condition should be consistency.
- I think we should use the `is_string` to validate the `$hint` is string before passing to the `preg_match`.

File: src/Container/Container.php (hunk 3 above)
Comment (peter279k) [reply src/Container/Container.php]: I've mentioned some tips about type check on previous comment.

Comment (lablnet) [reply src/Container/Container.php]: so we can check like
```php
if (class_exists($class)) {
        //Do stuff....
}
```

File: src/Contracts/Container/Container.php (hunk 5 above)
Comment (lablnet) [reply src/Contracts/Container/Container.php:69]: explain more please

File: src/Container/Container.php (hunk 3 above)
Comment (peter279k) [reply src/Container/Container.php]: I think the type hints validation should be another method.

And it's responsible for validating type.

File: src/Contracts/Container/Container.php (hunk 5 above)
Comment (peter279k) [reply src/Contracts/Container/Container.php:69]: It can use the `class_exists` to validate `$class`.

File: src/Container/Container.php (hunk 2 above)
Comment (Lablnet) [reply src/Container/Container.php]: Okay thanks for your response

```php
	protected function parseHint($hint): string
	{
		if (is_string($hint) && preg_match("/a-zA-Z/i", $hint)) {
			return $hint;
		} elseif(Arrays::isReallyArray($hint)) {
			list($hint, $alias) = $hint;

			$this->aliases[$alias] = $hint;
		} else {
			throw new \InvalidArgumentException ("The {$hint} parameter should be array or string ".gettype($hind)." given", 500);
		}

		return $hint;
	}
```

Comment (lablnet) [reply src/Container/Container.php]: @peter279k  Thanks for your response, finally it become
```php
	protected function parseHint($hint): string
	{
		if (is_string($hint) && preg_match("/a-zA-Z/i", $hint)) {
			return $hint;
		} elseif(Arrays::isReallyArray($hint)) {
			list($hint, $alias) = $hint;
			$this->aliases[$alias] = $hint;

			return $hint;
		}

		throw new \InvalidArgumentException ("The {$hint} parameter should be array or string ".gettype($hind)." given", 500);
	}
```

File: src/Container/Container.php (hunk 3 above)
Comment (lablnet) [reply src/Container/Container.php]: You mean it validate is type hint is string or array?

File: src/Container/Container.php (hunk 2 above)
Comment (peter279k) [reply src/Container/Container.php]: Fix typo: `.gettype($hind)`. It's `$hint` :).

File: src/Container/Container.php (hunk 4 above)
Comment (lablnet) [reply src/Container/Container.php]: could you explain more, i thought you mean we can append typeHint to parameter array

Comment (peter279k) [reply src/Container/Container.php]: Consider following code snippets:

```php
...
public function closureFactory(\Closure $factory, array $parameters)
...
```

File: src/Container/Container.php (hunk 3 above)
Comment (lablnet) [reply src/Container/Container.php]: About "class_exists" actually its an object not class 
like we register as follow
```php
$this->container = new Container();
$this->container->registerSingleton([Arrays::class, 'array'], new \Zest\Data\Arrays);
```
so we cant able either class exists or not

Comment (peter279k) [reply src/Container/Container.php]: How about using the `instanceof`?

Comment (peter279k) [reply src/Container/Container.php]: Maybe the `class_exists` is fine, too.

**Comment by lablnet** [comment]
@peter279k  Thanks you so much for your time to complete this class...

**Pull request merged**